x-pack/platform/plugins/shared/observability_ai_assistant/server/service/client/operators/add_anonymization_data.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { filter, map, takeLast, defaultIfEmpty, OperatorFunction } from 'rxjs';
+import type { Message } from '../../../../common/types';
+import {
+  StreamingChatResponseEventType,
+  type MessageAddEvent,
+  type StreamingChatResponseEvent,
+} from '../../../../common/conversation_complete';
+
+/**
+ * operator to process events with deanonymization data and format messages with deanonymizations.
+ *
+ * @param allMessages The combined messages to use as a fallback if no deanonymization data is found
+ * @returns An Observable that emits a single array of messages with deanonymization data added
+ */
+export function addAnonymizationData(
+  messages: Message[]
+): OperatorFunction<StreamingChatResponseEvent, Message[]> {
+  return (source$) => {
+    // Find the latest event with deanonymization data
+    return source$.pipe(
+      filter(
+        (event): event is MessageAddEvent =>
+          event.type === StreamingChatResponseEventType.MessageAdd &&
+          !!(event.deanonymized_input || event.deanonymized_output)
+      ),
+      takeLast(1),
```

Comment:
we should do this for every MessageAddEvent, but only take the input for the first one.

Reply:
@neptunian I've changed this to match on the message content - for any MessageAddEvent, the `messages` that served as input can be different from the messages that are the input to this function. We only emit deanonymizations for `content`, and if `content` does not match, the deanonymizations are likely invalid too. There is an edge-case here where the exact same user message will be matched to multiple deanonymizations, however I'm pretty sure in those cases the deanonymizations should produce the exact same result, given how we redact those fragments.